### PR TITLE
mon/OSDMonitor: more fixes for reweight_by_* command

### PR DESCRIPTION
**Pull request by xiexingguo**

### CODE REVIEW

**Comment by xiexingguo** [comment]
@liewegas

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -638,6 +638,9 @@ int OSDMonitor::reweight_by_utilization(int oload,
       // is a factor to take into account the original weights,
       // to represent e.g. differing storage capacities
       unsigned weight = osdmap.get_weight(p->first);
+      if (weight == 0) {
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: why not move this out of this if block and right before https://github.com/ceph/ceph/pull/8941/commits/3538fff7a58e7668f6876ec398c9ac92edd25b43#diff-0a5db46a44ae9900e226289a810f10e8R634 .

Comment (xiexingguo) [inline src/mon/OSDMonitor.cc]: @tchaikov Upgraded, thank you.

**Comment by tchaikov** [comment]
lgtm

File: src/mon/OSDMonitor.cc [hunk 2]
```diff
@@ -592,6 +592,11 @@ int OSDMonitor::reweight_by_utilization(int oload,
     std::pair<int, float> osd_util;
     osd_util.first = p->first;
     if (by_pg) {
+      if (pgs_by_osd[p->first] == 0) {
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: `pgs_by_osd[p->first]` could read out of the bound if the osd id is greater than the largest osd id serving the specified pools. so maybe we should check in this way:
```c++
if (p->first >= (int)pgs_by_osd.size()) {
```

**Comment by tchaikov** [comment]
withdraw my lgtm, as https://github.com/ceph/ceph/pull/8941#discussion_r62284023 needs to be addressed first.

File: src/mon/OSDMonitor.cc (hunk 2 above)
Comment (xiexingguo) [inline src/mon/OSDMonitor.cc]: Seems we have set the array size at max_osd at 
https://github.com/ceph/ceph/pull/8941/files#diff-0a5db46a44ae9900e226289a810f10e8R500.

File: src/mon/OSDMonitor.cc [hunk 3]
```diff
@@ -519,6 +519,10 @@ int OSDMonitor::reweight_by_utilization(int oload,
 	if (*q >= (int)pgs_by_osd.size())
 	  pgs_by_osd.resize(*q);
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: @xiexingguo then we can zap this line, or replace it with 
```c++
assert(*q < (int)pgs_by_osd.size());
```

File: src/mon/OSDMonitor.cc [hunk 4]
```diff
@@ -519,6 +519,10 @@ int OSDMonitor::reweight_by_utilization(int oload,
...
 	if (pgs_by_osd[*q] == 0) {
+          if (osdmap.crush->get_item_weightf(*q) <= 0) {
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: is it possible that we will try to increase the weight of a zero-weighted osd?

Comment (xiexingguo) [inline src/mon/OSDMonitor.cc]: @tchaikov  Upgraded https://github.com/ceph/ceph/pull/8941/commits/6effa004dd060f3e1c0a5e057d642a667f057d28 to a safer way.

Comment (xiexingguo) [inline src/mon/OSDMonitor.cc]: And I doubt we shall let this happen, otherwise we will encounter an **division by zero** error at https://github.com/ceph/ceph/pull/8941/files#diff-0a5db46a44ae9900e226289a810f10e8R611

File: src/mon/OSDMonitor.cc (hunk 1 above)
Comment (xiexingguo) [inline src/mon/OSDMonitor.cc]: oops, wait, I revert this change by accident...

**Comment by yuriw** [comment]
run http://pulpito.ceph.com/yuriw-2016-05-24_13:12:38-rados-wip-yuri-testing2-distro-basic-smithi/
re-run of failed http://pulpito.ceph.com/yuriw-2016-05-24_20:25:47-rados-wip-yuri-testing2-distro-basic-smithi/
